bycycle/features/shape.py

```diff
@@ -0,0 +1,213 @@
+        - ``time_rise`` : time between peak and previous trough
+        - ``time_peak`` : time between rise and decay zero-crosses
+        - ``time_trough`` : duration of previous trough estimated by zero-crossings
+        - ``volt_decay`` : voltage change between peak and next trough
+        - ``volt_rise`` : voltage change between peak and previous trough
+        - ``volt_amp`` : average of rise and decay voltage
+        - ``volt_peak`` : voltage at the peak
+        - ``volt_trough`` : voltage at the last trough
+        - ``time_rdsym`` : fraction of cycle in the rise period
+        - ``time_ptsym`` : fraction of cycle in the peak period
+        - ``band_amp`` : average analytic amplitude of the oscillation computed using narrowband
+          filtering and the Hilbert transform. Filter length is 3 cycles of the low cutoff
+          frequency. Average taken across all time points in the cycle.
+
+    df_samples : pandas.DataFrame, optional, default: True
+        Dataframe containing sample indices of cyclepoints.
+        Columns (listed for peak-centered cycles):
+
+        - ``sample_peak`` : sample of 'sig' at which the peak occurs
+        - ``sample_zerox_decay`` : sample of the decaying zero-crossing
+        - ``sample_zerox_rise`` : sample of the rising zero-crossing
+        - ``sample_last_trough`` : sample of the last trough
+        - ``sample_next_trough`` : sample of the next trough
+
+    Notes
+    -----
+    Peak vs trough centering
+
+        - By default, the first extrema analyzed will be a peak, and the final one a trough.
+        - In order to switch the preference, the signal is simply inverted and columns are renamed.
+        - Columns are slightly different depending on if ``center_extrema`` is set to 'peak' or
+          'trough'.
+
+    """
+
+    # Set defaults if user input is None
+    if find_extrema_kwargs is None:
+        find_extrema_kwargs = {'filter_kwargs': {'n_cycles': 3}}
+
+    elif 'first_extrema' in find_extrema_kwargs.keys():
+        raise ValueError('''This function has been designed to assume that the first extrema
+            identified will be a peak. This cannot be overwritten at this time.''')
+
+    # Negate signal if set to analyze trough-centered cycles
+    if center_extrema == 'peak':
+        pass
+    elif center_extrema == 'trough':
+        sig = -sig
+    else:
+        raise ValueError('Parameter "center_extrema" must be either "peak" or "trough"')
+
+    # Find extrema and zero-crossings locations in the signal
+    peaks, troughs = find_extrema(sig, fs, f_range, **find_extrema_kwargs)
+    rises, decays = find_zerox(sig, peaks, troughs)
+
+    # For each cycle, identify the sample of each extrema and zero-crossing
+    df_samples = compute_samples(peaks, troughs, decays, rises)
+
+    # Initialize dictionary for shape features
+    shape_features = {}
+
+    # Compute durations of period, peaks, and troughs
```

Comment:
These are all very small calculations, but it means this function is calculating a lot of different things overall, and there is no way to compute each measure independently. I think maybe some (all?) of these computation should be implemented as sub-functions, that operate on arrays, that get called from this function. 

These would be very simple array functions, doing quick calculations, but I think it would help with the modularity a bit. 
For example:
`def compute_distance(vals1, vals): return vals1 - vals2`
Could be used to compute durations, etc.

Reply:
I broke these down in to sub functions (f31f24c).

Reply:
I broke these down in to sub functions (f31f24c).